### Теперь займемся подготовкой спаршенного датасета
* Разберемся с пропусками
* Преобразуем имеющиеся признаки и выведем новые
* Сформируем финальное описание признаков датасета 

In [102]:
import pandas as pd
df_final = pd.read_csv("../../data/df_final.csv")

In [103]:
df_final

,steam_id,game_name,game_description_snippet,game_price,found_game_price,all_language_reviews_type,all_language_reviews_count,has_russian_reviews,all_russian_reviews_type,all_russian_reviews_count,...,igdb_id,type,first_release_date,rating,rating_count,total_rating,total_rating_count,popularity_igdb,popularity_steam,popularity_twitch
0,1313940,My Holiday 2,This is the second part of the long holiday of...,259 руб.,1.0,Очень положительные,106.0,0.0,NaN,NaN,...,150866,NaN,2020-10-05,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,1902870,Vertical Kingdom,Возродите былое величие Империи в этой карточн...,710 руб.,1.0,В основном положительные,111.0,0.0,NaN,NaN,...,202602,"GOG, Nintendo",2024-04-15,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,1071940,Streamer's Life,"Игра в жанре RPG-Simulation, в которой вам пре...",NaN,0.0,В основном положительные,139.0,0.0,NaN,NaN,...,124447,NaN,2019-07-31,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,1706570,Chupacabras: Night Hunt,Hunt the chupacabras alone or with your friend...,99 руб.,1.0,В основном положительные,124.0,0.0,NaN,NaN,...,163904,NaN,2021-08-20,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,1085150,Golf Defied,"Физика и механики нового поколения, мультиплее...",Бесплатные,1.0,Смешанные,175.0,0.0,NaN,NaN,...,118634,NaN,2019-03-11,NaN,NaN,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
25734,1547900,Active Mummy,"Active Mummy — динамичный 2D-платформер, в кот...",210 руб.,1.0,Положительные,20.0,0.0,NaN,NaN,...,156747,NaN,2021-02-28,NaN,NaN,NaN,NaN,NaN,NaN,NaN
25735,2238360,You Have No Time,Rev up your inner speedster in this high-octan...,165 руб.,1.0,Положительные,13.0,0.0,NaN,NaN,...,244854,NaN,2023-11-09,NaN,NaN,NaN,NaN,NaN,NaN,NaN
25736,1490030,Cyber Dodge,Cyber Dodge is a skill-based Runner and a Die ...,61 руб.,1.0,Обзоров пользователей: 3,3.0,0.0,NaN,NaN,...,159989,NaN,2021-02-14,NaN,NaN,NaN,NaN,NaN,NaN,NaN
25737,1855190,Vektor Z,Vektor Z brings fast and furious retro space b...,280 руб.,1.0,Обзоров пользователей: 2,2.0,0.0,NaN,NaN,...,186672,NaN,2022-01-14,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [104]:
df_final.info()

<class 'pandas.DataFrame'>
RangeIndex: 25739 entries, 0 to 25738
Data columns (total 22 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   steam_id                    25739 non-null  int64  
 1   game_name                   25739 non-null  str    
 2   game_description_snippet    25217 non-null  str    
 3   game_price                  19831 non-null  str    
 4   found_game_price            24785 non-null  float64
 5   all_language_reviews_type   24418 non-null  str    
 6   all_language_reviews_count  24418 non-null  float64
 7   has_russian_reviews         24785 non-null  float64
 8   all_russian_reviews_type    1543 non-null   str    
 9   all_russian_reviews_count   1543 non-null   float64
 10  steam_app_url               24785 non-null  str    
 11  support_ru_region           24785 non-null  float64
 12  igdb_id                     25739 non-null  int64  
 13  type                        12909 non-null

## Посмотрим на неудачно спаршенные игры

Посмотрев на вывод .info() по датасету можно заметить, что у нас есть ряд признаков, где у всех non-null значений по 24785. В частности, это бинарные T/F признаки found_game_price, has_russian_reviews, support_ru_region. То есть парсер не то что не смог найти значения или нет, он не смог определить/ответить есть они или нет.

In [105]:
print(f'{df_final['found_game_price'].isna().sum() / df_final.shape[0] * 100}%')

3.7064377015424066%


И т.к. этих пропусков всего 3.7%, то можно их просто удалить

In [106]:
df_final = df_final.dropna(subset=['found_game_price'])
df_final.info()

<class 'pandas.DataFrame'>
Index: 24785 entries, 0 to 25738
Data columns (total 22 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   steam_id                    24785 non-null  int64  
 1   game_name                   24785 non-null  str    
 2   game_description_snippet    24764 non-null  str    
 3   game_price                  19831 non-null  str    
 4   found_game_price            24785 non-null  float64
 5   all_language_reviews_type   24418 non-null  str    
 6   all_language_reviews_count  24418 non-null  float64
 7   has_russian_reviews         24785 non-null  float64
 8   all_russian_reviews_type    1543 non-null   str    
 9   all_russian_reviews_count   1543 non-null   float64
 10  steam_app_url               24785 non-null  str    
 11  support_ru_region           24785 non-null  float64
 12  igdb_id                     24785 non-null  int64  
 13  type                        12402 non-null  str

Теперь у нас количество non-null значений в тех столбцах равно количеству строк в датасете

## Начнем с обработки колонки с ценой

Если у нас была найдена цена, то в рублях и тогда формат записи цены будет, к примеру: 165 руб.

Проверим, действительно ли везде рубли и можем ли мы тогда при наличии более 1 элемента после разделения по пробелу - удалить это обозначение валюты

In [107]:
df_final['game_price'].str.split(' ').str[1].value_counts()

game_price
руб.           16966
to                50
To                 4
WARMACHINE:        1
Name: count, dtype: int64

Посмотрим что за 55 строк с "to", "To" и "WARMACHINE"

In [108]:
df_final[df_final['game_price'].str.contains('to')]

,steam_id,game_name,game_description_snippet,game_price,found_game_price,all_language_reviews_type,all_language_reviews_count,has_russian_reviews,all_russian_reviews_type,all_russian_reviews_count,...,igdb_id,type,first_release_date,rating,rating_count,total_rating,total_rating_count,popularity_igdb,popularity_steam,popularity_twitch
988,299420,CubeGun,A Sandbox Multiplayer Airship Shooting Game.De...,Free to Play,1.0,Смешанные,321.0,0.0,NaN,NaN,...,17372,NaN,2014-05-21,50.000000,0.0,50.000000,0.0,7.848961e-07,0.000010,NaN
1049,266030,DRAKERZ-Confrontation,The Trading Card Game in Augmented Reality.,Free to Play,1.0,Смешанные,259.0,0.0,NaN,NaN,...,16876,NaN,2014-05-19,NaN,NaN,NaN,NaN,NaN,0.000009,NaN
1185,365050,Escape,Escape is a simple 2D game with 35 levels alre...,Free to Play,1.0,Смешанные,290.0,0.0,NaN,NaN,...,35380,NaN,2015-04-20,NaN,NaN,NaN,NaN,NaN,NaN,NaN
8678,250420,8BitMMO,8BitMMO is a retro-style 2D massively multipla...,Free to Play,1.0,Смешанные,4321.0,1.0,Смешанные,350.0,...,30232,NaN,2015-01-26,50.000000,2.0,50.000000,2.0,9.418753e-06,0.000137,NaN
8682,365270,Shadow Hunter,Shadow Hunter is a casual deductive puzzle gam...,Free to Play,1.0,Смешанные,73.0,0.0,NaN,NaN,...,35386,Steam,2015-04-20,NaN,NaN,NaN,NaN,NaN,NaN,NaN
10357,201210,Pox Nora,Delve into classic turn-based strategy gaming ...,Free to Play,1.0,В основном положительные,882.0,0.0,NaN,NaN,...,9595,NaN,2004-08-01,NaN,NaN,NaN,NaN,2.653102e-05,0.000025,NaN
10687,340810,Deepworld,Deepworld is a massively-multiplayer 2D crafti...,Free to Play,1.0,Смешанные,737.0,0.0,NaN,NaN,...,35979,NaN,2015-04-22,NaN,NaN,NaN,NaN,NaN,NaN,NaN
10781,360870,Ultimate Tic-Tac-Toe,Ultimate Tic-Tac-Toe takes the classic game of...,Free to Play,1.0,Очень положительные,1542.0,0.0,NaN,NaN,...,35501,NaN,2015-04-28,NaN,NaN,NaN,NaN,NaN,NaN,NaN
12026,343740,InMind VR,InMind is a short scientific adventure with ar...,Free to Play,1.0,В основном положительные,432.0,0.0,NaN,NaN,...,19024,NaN,2015-01-19,NaN,NaN,NaN,NaN,NaN,NaN,NaN
12047,361630,404Sight,404Sight is a free game made in support of net...,Free to Play,1.0,Очень положительные,1761.0,0.0,NaN,NaN,...,19389,NaN,2015-04-16,59.847319,6.0,59.847319,6.0,NaN,NaN,NaN


In [109]:
df_final[df_final['game_price'].str.contains('To')]

,steam_id,game_name,game_description_snippet,game_price,found_game_price,all_language_reviews_type,all_language_reviews_count,has_russian_reviews,all_russian_reviews_type,all_russian_reviews_count,...,igdb_id,type,first_release_date,rating,rating_count,total_rating,total_rating_count,popularity_igdb,popularity_steam,popularity_twitch
5966,348840,UFHO2,Play an alien navigating a maze of hexagons in...,Free To Play,1.0,Положительные,38.0,0.0,NaN,NaN,...,35793,NaN,2014-04-01,NaN,NaN,NaN,NaN,NaN,NaN,NaN
19475,298260,Only If,Only If is a surreal first person adventure-pu...,Free To Play,1.0,Смешанные,5954.0,1.0,Смешанные,362.0,...,17354,NaN,2014-07-25,69.525071,18.0,74.762536,19.0,NaN,NaN,NaN
19531,263540,Villagers and Heroes,Лучшая 3D MMO-игра! Фэнтезийная RPG со сражени...,Free To Play,1.0,Смешанные,3428.0,1.0,Смешанные,437.0,...,16843,NaN,2014-03-04,NaN,NaN,NaN,NaN,NaN,NaN,NaN
21807,290730,My Lands: Black Gem Hunting,My Lands – бесплатная реалтаймовая онлайн-стра...,Free To Play,1.0,Очень положительные,3308.0,1.0,Очень положительные,1875.0,...,10576,NaN,2014-09-16,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [110]:
df_final[df_final['game_price'].str.contains('WARMACHINE:')]

,steam_id,game_name,game_description_snippet,game_price,found_game_price,all_language_reviews_type,all_language_reviews_count,has_russian_reviews,all_russian_reviews_type,all_russian_reviews_count,...,igdb_id,type,first_release_date,rating,rating_count,total_rating,total_rating_count,popularity_igdb,popularity_steam,popularity_twitch
24973,253510,WARMACHINE: Tactics,Turn-based tactical combat set in the award-wi...,Play WARMACHINE: Tactics Demo,1.0,Смешанные,1160.0,0.0,NaN,NaN,...,16695,NaN,2014-10-04,50.0,1.0,50.0,1.0,0.000013,0.000044,5.308210e-07


Вышло, что первые два случая это бесплатные игры, а у одной игры неверно нашлась цена, но видно что там есть демо версия, а значит платная версия.

Для этой одной игры можно было не полениться и посмотреть вручную цену - 419 руб. Заменим пока у нее цену, а также посмотрим как еще обозначаются бесплатные игры

In [111]:
df_final.loc[24973, 'game_price'] = '419 руб.' # первое это индекс строки, второе уже имя столбца
df_final[df_final['steam_id'] == 253510]

,steam_id,game_name,game_description_snippet,game_price,found_game_price,all_language_reviews_type,all_language_reviews_count,has_russian_reviews,all_russian_reviews_type,all_russian_reviews_count,...,igdb_id,type,first_release_date,rating,rating_count,total_rating,total_rating_count,popularity_igdb,popularity_steam,popularity_twitch
24973,253510,WARMACHINE: Tactics,Turn-based tactical combat set in the award-wi...,419 руб.,1.0,Смешанные,1160.0,0.0,NaN,NaN,...,16695,NaN,2014-10-04,50.0,1.0,50.0,1.0,0.000013,0.000044,5.308210e-07


In [112]:
first_el_price = df_final['game_price'].str.split(' ').str[0].reset_index()
first_el_price['game_price'].value_counts()

game_price
Бесплатно     1966
42            1398
82            1236
200           1113
Бесплатные     844
              ... 
346,85           1
977              1
71               1
3499             1
1020             1
Name: count, Length: 498, dtype: int64

Попробуем преобразовать все значения в числа, что не сможем - выведем

Также заметим, что нам здесь уже видно дробное число с зяпятой вместо точки для разделения целой и дробной частей. Заменим для начала запятые на точки. Сделаем это в исходном датасете 

In [113]:
df_final['game_price'] = df_final['game_price'].str.replace(',', '.')

In [114]:
first_el_price = df_final['game_price'].str.split(' ').str[0].reset_index()
# https://stackoverflow.com/questions/73313363/how-to-deal-with-strings-on-a-numeric-column-in-pandas

first_el_price['game_price_nums'] = pd.to_numeric(first_el_price['game_price'], errors='coerce')

first_el_price.loc[first_el_price['game_price'].notna() & first_el_price['game_price_nums'].isna(), 'game_price'].value_counts()

game_price
Бесплатно     1966
Бесплатные     844
Free            54
Name: count, dtype: int64

Мы нашли три варианта обозначения бесплатной игры, где "Free" как первый элемент это для строк, которые были найдены раньше, где значения "Free to Play" и "Free To Play"

Заменим их все на 0, а также уберем "руб."

In [115]:
df_final['game_price'] = df_final['game_price'].str.replace('Бесплатно', '0')
df_final['game_price'] = df_final['game_price'].str.replace('Бесплатные', '0')
df_final['game_price'] = df_final['game_price'].str.replace('Free to Play', '0')
df_final['game_price'] = df_final['game_price'].str.replace('Free To Play', '0')
df_final['game_price'] = df_final['game_price'].str.replace('руб.', '')

Попробуем теперь применить pd.to_numeric уже в исходном df и без указания errors='coerce'. Если все пройдет без ошибок, знвчит все значения числовые, если будет ошибка то продолжим разбираться

In [116]:
df_final['game_price'] = pd.to_numeric(df_final['game_price'])

In [117]:
df_final['game_price'].dtype

dtype('float64')

Отлично, все значения преобразовались в число, в данном случае float

Проверим также верно ли указано наличие/отсутствие данных о цене в признаке found_game_price

In [118]:
not_found_game_price = df_final[df_final['found_game_price'] == 0]
not_found_game_price[not_found_game_price['game_price'].notna()]

,steam_id,game_name,game_description_snippet,game_price,found_game_price,all_language_reviews_type,all_language_reviews_count,has_russian_reviews,all_russian_reviews_type,all_russian_reviews_count,...,igdb_id,type,first_release_date,rating,rating_count,total_rating,total_rating_count,popularity_igdb,popularity_steam,popularity_twitch


In [119]:
not_found_game_price = df_final[df_final['found_game_price'] == 1]
not_found_game_price[not_found_game_price['game_price'].isna()]

,steam_id,game_name,game_description_snippet,game_price,found_game_price,all_language_reviews_type,all_language_reviews_count,has_russian_reviews,all_russian_reviews_type,all_russian_reviews_count,...,igdb_id,type,first_release_date,rating,rating_count,total_rating,total_rating_count,popularity_igdb,popularity_steam,popularity_twitch


Для тех строк, где указано что цена найдена - пропусков нет. Где не найдена - все пропуски. Делаем вывод что признак определился верно на этапе скрапинга.